pkg/action/install.go

```diff
@@ -1,43 +1,164 @@
+func (i *Install) SetNamespace(chart *chart.Chart, defaultns string) {
+	i.Namespace = defaultns
+	if chart.Metadata.Annotations != nil {
+		if val, ok := chart.Metadata.Annotations["hypper.cattle.io/namespace"]; ok {
+			i.Namespace = val
+		} else {
+			if val, ok := chart.Metadata.Annotations["catalog.cattle.io/namespace"]; ok {
+				i.Namespace = val
 			}
 		}
-		missing = append(missing, r.Name)
+	}
+}
+
+// Name returns the name that should be used.
+//
+// This will read the flags and handle name generation if necessary.
+func (i *Install) Name(chart *chart.Chart, args []string) (string, error) {
+	// args here will only be: [NAME] [CHART]
+	// cobra flags have been already stripped
+
+	flagsNotSet := func() error {
+		if i.GenerateName {
+			return errors.New("cannot set --generate-name and also specify a name")
+		}
+		if i.NameTemplate != "" {
+			return errors.New("cannot set --name-template and also specify a name")
+		}
+		return nil
 	}
 
-	if len(missing) > 0 {
-		return errors.Errorf("found in Chart.yaml, but missing in charts/ directory: %s", strings.Join(missing, ", "))
+	if len(args) > 2 {
+		return args[0], errors.Errorf("expected at most two arguments, unexpected arguments: %v", strings.Join(args[2:], ", "))
 	}
-	return nil
+
+	if len(args) == 2 {
+		return args[0], flagsNotSet()
+	}
+
+	if chart.Metadata.Annotations != nil {
+		if val, ok := chart.Metadata.Annotations["hypper.cattle.io/release-name"]; ok {
+			return val, nil
+		}
+		if val, ok := chart.Metadata.Annotations["catalog.cattle.io/release-name"]; ok {
+			return val, nil
+		}
+	}
+
+	if i.NameTemplate != "" {
+		name, err := action.TemplateName(i.NameTemplate)
+		return name, err
+	}
+
+	if i.ReleaseName != "" {
+		return i.ReleaseName, nil
+	}
+
+	if !i.GenerateName {
+		return "", errors.New("must either provide a name, set the correct chart annotations, or specify --generate-name")
+	}
+
+	base := filepath.Base(args[0])
+	if base == "." || base == "" {
+		base = "chart"
+	}
+	// if present, strip out the file extension from the name
+	if idx := strings.Index(base, "."); idx != -1 {
+		base = base[0:idx]
+	}
+
+	return fmt.Sprintf("%s-%d", base, time.Now().Unix()), nil
+}
+
+// Chart returns the chart that should be used.
+//
+// This will read the flags and skip args if necessary.
+func (i *Install) Chart(args []string) (string, error) {
+	if len(args) > 2 {
+		return args[1], errors.Errorf("expected at most two arguments, unexpected arguments: %v", strings.Join(args[2:], ", "))
+	}
+
+	if len(args) == 2 {
+		return args[1], nil
+	}
+
+	// len(args) == 1
+	return args[0], nil
+}
+
+// NameAndChart overloads Helm's NameAndChart. It always fails.
+//
+// On Hypper, we need to read the chart annotations to know the correct release name.
+// Therefore, it cannot happen in this function.
+func (i *Install) NameAndChart(args []string) (string, string, error) {
+	return "", "", errors.New("NameAndChart() cannot be used")
```

Comment:
why not return

```golang
return nil, nil, errors.New("NameAndChart() cannot be used")
```

I do not see the point of having empty strings here since the error should be evaluated anyway in the code that uses this function. Also empty string might be valid when accidentely used and can cause follow-up issues.

Reply:
well, the default value for a string is `""`, not `nil`.  The idea of this function overload is to fail always to deactivate this function. This is one function that we obtained from Helm's type composition, but it can't return correct values, as in Hypper we need to look at the chart annotations, and the signature of this function doesn't get a chart.

Reply:
Yes, but if you return an error anyway, returning nil for the expected values is often used. So I'd prefer that.

Reply:
There are no nil strings in go, so I don't think one can do what you suggest. See for Helm for example:
https://github.com/helm/helm/blob/b30f61d76de3805a839a95dc0f8ef8d8e208b09f/pkg/action/install.go#L564
https://github.com/helm/helm/blob/b30f61d76de3805a839a95dc0f8ef8d8e208b09f/pkg/action/install.go#L638
and others.